File: backend/app/repositories/rag_documents.py
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class RagDocumentsRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    def match_rag_documents(
        self,
        *,
        query_embedding: list[float],
        match_count: int,
        filter_ticker: str | None,
        filter_document_type: str | None,
    ) -> list[dict[str, Any]]:
        payload: dict[str, Any] = {
            "query_embedding": query_embedding,
            "match_count": int(match_count),
            "filter_ticker": filter_ticker,
            "filter_document_type": filter_document_type,
        }

        try:
            resp = self._supa.rpc("match_rag_documents", payload).execute()
        except Exception as exc:
            msg = str(exc)
            if "different vector dimensions" in msg or "vector dimensions" in msg:
                logger.error(
                    "Supabase RPC match_rag_documents failed due to embedding dimension mismatch. "
                    "query_dim=%s. This usually means rag_documents contains mixed embedding dimensions. "
                    "Fix by updating the SQL function to filter `d.dimension = vector_dims(query_embedding)`.",
                    len(query_embedding),
                )
            raise

        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

File: backend/internal/repo/rag_documents.go
package repo

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"yunews/backend/internal/supabase"
)

// RagDocumentsRepository handles the match_rag_documents RPC function.
type RagDocumentsRepository struct {
	db *supabase.Client
}

// NewRagDocumentsRepository creates a new RagDocumentsRepository.
func NewRagDocumentsRepository(db *supabase.Client) *RagDocumentsRepository {
	return &RagDocumentsRepository{db: db}
}

// MatchRagDocuments executes the match_rag_documents Supabase RPC to perform vector search.
func (r *RagDocumentsRepository) MatchRagDocuments(
	queryEmbedding []float32,
	matchCount int,
	filterTicker *string,
	filterDocumentType *string,
) ([]map[string]interface{}, error) {
	params := map[string]interface{}{
		"query_embedding":     queryEmbedding,
		"match_count":         matchCount,
		"filter_ticker":       filterTicker,
		"filter_document_type": filterDocumentType,
	}

	data, err := r.db.Rpc("match_rag_documents", params)
	if err != nil {
		msg := err.Error()
		if strings.Contains(msg, "different vector dimensions") || strings.Contains(msg, "vector dimensions") {
			slog.Error("match_rag_documents: embedding dimension mismatch",
				"query_dim", len(queryEmbedding),
				"hint", "Fix by updating the SQL function to filter d.dimension = vector_dims(query_embedding)")
		}
		return nil, fmt.Errorf("match_rag_documents rpc: %w", err)
	}

	var rows []map[string]interface{}
	if err := json.Unmarshal(data, &rows); err != nil {
		return nil, fmt.Errorf("parse match_rag_documents response: %w", err)
	}
	return rows, nil
}
